ext/LinearSolveHYPRE.jl

```diff
@@ -0,0 +1,193 @@
+module LinearSolveHYPRE
+
+using HYPRE.LibHYPRE: HYPRE_Complex
+using HYPRE: HYPRE, HYPREMatrix, HYPRESolver, HYPREVector
+using IterativeSolvers: Identity
+using LinearSolve: HYPREAlgorithm, LinearCache, LinearProblem, LinearSolve,
+                   OperatorAssumptions, default_tol, init_cacheval, issquare, set_cacheval
+using SciMLBase: LinearProblem, SciMLBase
+using UnPack: @unpack
+
+mutable struct HYPRECache
+    solver::Union{HYPRE.HYPRESolver, Nothing}
+    A::Union{HYPREMatrix, Nothing}
+    b::Union{HYPREVector, Nothing}
+    oidb::UInt # objectid(b) to keep track of isfresh for b
+    u::Union{HYPREVector, Nothing}
+end
+
+function LinearSolve.init_cacheval(alg::HYPREAlgorithm, A, b, u, Pl, Pr, maxiters::Int,
+                                   abstol, reltol,
+                                   verbose::Bool, assumptions::OperatorAssumptions)
+    return HYPRECache(nothing, nothing, nothing, UInt(0), nothing)
+end
+
+# Note:
+# SciMLBase.init is overloaded here instead of just LinearSolve.init_cacheval for two
+# reasons:
+# - HYPREArrays can't really be `deepcopy`d, so that is turned off by default
+# - The solution vector/initial guess u0 can't be created with
+#   fill!(similar(b, size(A, 2)), false) since HYPREArrays are not AbstractArrays.
```

Comment:
Could be an idea to define an internal function for this that take A and b and allocate a appropriate solution vector. (Won't help in this case due to the copy issue though).

Reply:
That's what `alias_A` and `alias_b` are there to handle, though we should probably make it on demand as a solver trait, i.e. if `alias_A = false` but `solver_doesnt_alias_A(alg)`, then still don't copy. We need to do safety by default since otherwise it confuses users (always better to safety by default, and then opt in to remove the allocation), we had quite a few issues opened about this before. That said, I agree that if the solver doesn't modify `A` anyways, then it doesn't need to deepcopy it. So we can keep this init overload for now and do a trait-based fix as a separate thing.

Reply:
Yea, thats what I meant, it should only be used as a hint to the alg whether or not to copy.

Reply:
It can probably just be a copy.